### PR TITLE
jit(arm64): support for memory size/grow, host functions

### PR DESCRIPTION
**Pull request by mathetake**
part of #187

### CODE REVIEW

**Comment by mathetake** [comment]
ok not time to polish

**Comment by mathetake** [comment]
now

**Review by nullpo-head — APPROVED** [review]
LGTM. Thanks!

File: wasm/jit/jit_amd64.go [hunk 1]
```diff
@@ -175,23 +175,6 @@ func (c *amd64Compiler) compileHostFunction(address wasm.FunctionAddress) error
 	if err := c.callGoFunction(jitCallStatusCodeCallHostFunction, address); err != nil {
 		return err
 	}
-
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:178]: PR comment why we aren't looking at this anymore?

File: wasm/jit/jit_amd64.go [hunk 2]
```diff
@@ -4241,7 +4224,7 @@ func (c *amd64Compiler) compileMemorySize() error {
 	getMemorySizeInPageUnit.To.Type = obj.TYPE_REG
 	getMemorySizeInPageUnit.To.Reg = loc.register
 	getMemorySizeInPageUnit.From.Type = obj.TYPE_CONST
-	getMemorySizeInPageUnit.From.Offset = 16
+	getMemorySizeInPageUnit.From.Offset = wasm.MemoryPageSizeInBit
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:4227]:
```suggestion
	getMemorySizeInPageUnit.From.Offset = wasm.MemoryPageSizeInBits
```

File: wasm/jit/jit_arm64.go [hunk 3]
```diff
@@ -278,8 +286,13 @@ func (c *arm64Compiler) compileTwoRegistersToNoneInstruction(instruction obj.As,
 }
 
 func (c *arm64Compiler) compileUnconditionalBranchInstruction() (br *obj.Prog) {
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64.go:288]: maybe inline now?

File: wasm/jit/jit_arm64.go [hunk 4]
```diff
@@ -334,18 +347,19 @@ func (c *arm64Compiler) pushFunctionParams() {
...
 	c.compileNOP()
 
 	c.pushFunctionParams()
 
-	// TODO: Add maybeGrowValueStack() .
-
-	// Before executing function body, we must initialize the stack base pointer register
-	// so that we can manipulate the memory stack properly.
-	if err := c.compileReservedStackBasePointerRegisterInitialization(); err != nil {
+	// Check if it's necessary to grow the value stack before entring function body.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64.go:355]:
```suggestion
	// Check if it's necessary to grow the value stack before entering function body.
```

File: wasm/jit/jit_arm64.go [hunk 5]
```diff
@@ -354,6 +368,63 @@ func (c *arm64Compiler) compilePreamble() error {
 	return nil
 }
 
+// compileMaybeGrowValueStack adds instructions to check the necessity to grow the value stack,
+// and if so, make the builtin function call to do so. These instructions are called in the function's
+// preamble.
+func (c *arm64Compiler) compileMaybeGrowValueStack() error {
+	tmpRegs, found := c.locationStack.takeFreeRegisters(generalPurposeRegisterTypeInt, 3)
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64.go:375]:
```suggestion
	tmpRegs, found := c.locationStack.takeFreeRegisters(generalPurposeRegisterTypeInt, 2)
```

File: wasm/jit/jit_arm64.go [hunk 6]
```diff
@@ -474,8 +542,15 @@ func (c *arm64Compiler) exit(status jitCallStatusCode) error {
 	return nil
 }
 
+// compileHostFunction implements compiler.compileHostFunction for the arm64 architecture.
 func (c *arm64Compiler) compileHostFunction(address wasm.FunctionAddress) error {
-	return errors.New("TODO: implement compileHostFunction on arm64")
+	// First we must update the location stack to reflect the number of host function inputs.
+	c.pushFunctionParams()
+
+	if err := c.compileCallGoFunction(jitCallStatusCodeCallHostFunction, address); err != nil {
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64.go:550]: note: I like the name of this

File: wasm/jit/jit_arm64.go [hunk 7]
```diff
@@ -948,14 +1033,29 @@ func (c *arm64Compiler) compileCallFunction(addr wasm.FunctionAddress, functype
 	return nil
 }
 
-// readInstructionAddress adds an ADR instruction to set the absolute address of "target instruction"
+// compileCalcCallFrameStackTopAddress adds instruction to set the absolute address of
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64.go:1036]:
```suggestion
// compileCalcCallFrameStackTopAddress adds instructions to set the absolute address of
```

File: wasm/jit/jit_arm64.go [hunk 8]
```diff
@@ -2265,7 +2454,6 @@ func (c *arm64Compiler) compileReservedStackBasePointerRegisterInitialization()
 	c.compileAddInstructionWithLeftShiftedRegister(
 		reservedRegisterForTemporary, 3, reservedRegisterForStackBasePointerAddress,
 		reservedRegisterForStackBasePointerAddress)
-	return nil
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64.go:2268]: thanks

File: wasm/jit/jit_arm64.go [hunk 9]
```diff
@@ -2301,9 +2489,7 @@ func (c *arm64Compiler) compileModuleContextInitialization() error {
 
 	// If the module instance address stays the same, we could skip the entire code below.
 	c.compileTwoRegistersToNoneInstruction(arm64.ACMP, moduleInstanceAddressRegister, tmpX)
-	brIfMdouleUnchanged := c.newProg()
-	brIfMdouleUnchanged.As = arm64.ABEQ
-	brIfMdouleUnchanged.To.Type = obj.TYPE_BRANCH
+	brIfMdouleUnchanged := c.compilelBranchInstruction(arm64.ABEQ)
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64.go:2492]: s/Mdoule/Module/g

File: wasm/jit/jit_arm64_test.go [hunk 10]
```diff
@@ -1673,79 +1668,95 @@ func TestArm64Compiler_readInstructionAddress(t *testing.T) {
...
+			// Check status and returned values.
+			require.Equal(t, jitCallStatusCodeReturned, env.jitStatus())
+			require.Equal(t, uint64(2), env.stackPointer()) // Must be 2 (dummy value + the calculation results)
+			require.Equal(t, uint64(0), env.stackBasePointer())
+			require.Equal(t, expectedValue, env.stackTopAsUint32())
+		})
+	}
 }
 
 func TestArm64Compiler_compieleSelect(t *testing.T) {
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64_test.go:1762]: s/compiele/compile/g

File: wasm/jit/jit_arm64_test.go [hunk 11]
```diff
@@ -2439,3 +2449,164 @@ func TestArm64Compiler_compileLoad(t *testing.T) {
 		})
 	}
 }
+
+func TestArm64Compiler_compileMemryGrow(t *testing.T) {
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64_test.go:2453]: s/Memry/Memory/g

File: wasm/jit/jit_arm64_test.go [hunk 12]
```diff
@@ -2439,3 +2449,164 @@ func TestArm64Compiler_compileLoad(t *testing.T) {
...
+		returnAddress := env.callFrameStackPeek().returnAddress
+		require.NotZero(t, returnAddress)
+		jitcall(returnAddress, uintptr(unsafe.Pointer(env.engine())))
+
+		// Check the result. This should be "Returned".
+		require.Equal(t, jitCallStatusCodeReturned, env.jitStatus())
+	})
+}
+
+func TestArm64Compiler_compieleHostFunction(t *testing.T) {
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64_test.go:2586]: s/compiele/compile/g

File: wasm/jit/jit_arm64.go [hunk 13]
```diff
@@ -474,8 +542,15 @@ func (c *arm64Compiler) exit(status jitCallStatusCode) error {
 	return nil
 }
 
+// compileHostFunction implements compiler.compileHostFunction for the arm64 architecture.
 func (c *arm64Compiler) compileHostFunction(address wasm.FunctionAddress) error {
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64.go:546]: should this be compileCallHostFunction?

File: wasm/jit/jit_arm64_test.go [hunk 14]
```diff
@@ -2439,3 +2449,164 @@ func TestArm64Compiler_compileLoad(t *testing.T) {
...
+	env := newJITEnvironment()
+	compiler := env.requireNewCompiler(t)
+
+	// The assembler skips the first instruction so we intentionally add NOP here.
+	// TODO: delete after #233
+	compiler.compileNOP()
+
+	addr := wasm.FunctionAddress(100)
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64_test.go:2594]: is it too much to actually call a go function?

**Review by codefromthecrypt — APPROVED** [review]
getting near 🏁

File: wasm/jit/jit_arm64_test.go (hunk 14 above)
Comment (mathetake) [reply wasm/jit/jit_arm64_test.go:2594]: too much and doesn't affect anything in the test, so I will leave as-is

File: wasm/jit/jit_amd64.go (hunk 1 above)
Comment (mathetake) [reply wasm/jit/jit_amd64.go:178]: this was accidentally added so I just deleted!

**Comment by mathetake** [comment]
thanks for the look!

**Pull request merged**